src/messages/MClientRequest.h

```diff
@@ -165,11 +165,10 @@ class MClientRequest : public Message {
     if (header.version >= 4) {
       ::decode(head, p);
     } else {
-      struct ceph_mds_request_head_legacy *old_mds_head =
-	(struct ceph_mds_request_head_legacy *)&(head.oldest_client_tid);
-
-      ::decode(*old_mds_head, p);
+      struct ceph_mds_request_head_legacy old_mds_head;
 
+      ::decode(old_mds_head, p);
+      memcpy(&head.oldest_client_tid, &old_mds_head, sizeof(old_mds_head));
```

Comment:
I think this is getting lost in translation. Enjoy your ability to use C++ and write "invisible" functions; instead of memcpy do something like
```head.copy_from_legacy(old_mds_head);```
or perhaps ```copy_legacy_head(head, old_mds_head)``` since these are declared as shared C structs. That way if we update stuff, it requires changing a single function located next to the ceph_mds_request_head struct instead of searching out everybody who copies them.

Granted, if we only have the one user maybe it's not such a big deal and is fine as it stands. That's just the sort of protective programming we can engage in even though it makes the kernel devs angry for being inefficient. ;)